### DISCUSSION TITLE
Benvenuto/a nelle Discussioni di LootBot!

### DISCUSSION THREAD

**Discussion by sidelux** [sidelux/LootBot#1] [Announcements]
## 👋 Benvenuto/a!
E' possibile tramite Discussioni interagire con gli altri membri della community.
* Fai domande generiche sul funzionamento del bot
* Condividi idee e suggerimenti per nuove funzionalità
* Non segnalare problemi, per quelli c'è l'apposita sezione Issues

Per iniziare, commenta sotto presentandoti e raccontandoci qualcosa.

**Comment by furins** [comment]
Ciao a tutti, e grazie @sidelux per aver reso open source il bot e averci dato la possibilità di contribuire e -spero- di alleggerirti almeno di alcuni aspetti più noiosi.
Sviluppo professionalmente da più 20 anni, principalmente backend e soprattutto in python (ma ho sviluppato in JS dal 2001 ad oggi, seppur controvoglia 🤣 ), gioco a Loot dal 2016, penso di poter dare una mano con il refactoring per modularizzare il codice  (senza introdurre nuove funzionalità e mantenendo di fatto il codice di Edo all'interno dei singoli moduli). 
A una prima scorsa del codice mi verrebbe da proporre una struttura del codice come segue:

- **src** per separare il codice dal resto
    - **bots** interazione con telegram (Loot, Plus)
    - **configuration** (ospiterebbe config.js)
    - **admin** metodi riservati all'amministratore del gioco
    - **logging** metodi per formattazione messaggi in console
    - **game** metodi per la gestione delle dinamiche di gioco
    - **export** metodi per l'output in formato pdf, ad esempio
    - **suggestions** (il codice attualmente è stato rimosso, pare, ma vi si fa riferimento nel codice dei bot)
    - **maintenance** comandi per le fasi di manutenzione
    -  ... probabilmente altri
- **test** (a mio avviso fondamentali prima di poter iniziare ad accettare PR)
- **docs** (non strettamente necessario all'inizio perchè può trovare spazio nelle sezioni discussion o wiki di github ma...  non sarebbe male se accompagnassero il codice)
- **migrations** (per preservare eventuali future migrazioni dello schema del DB)

sto iniziando a implementarla sul mio fork per permettervi di valutare se sembra una struttura funzionale. Fatemi sapere cosa ne pensate!

**Reply by sidelux** [reply]
È una delle modifiche che mi sarebbe piaciuto fare in modo abbastanza ordinato e dettagliato sì.
Il suggestion è un modulo separato che al momento non ho il permesso di integrare dato che non l'ho realizzato io, ma essendo appunto separato non c'è altro lavoro da farci, la piccola parte che si trova nel codice attuale si può farlo rientrare in un "function" o "extra".
Attendo il fork allora ;)

**Reply by furins** [reply]
Ciao @sidelux perdona la latitanza: avevo provato un paio di volte a preparare una PR ma le dimensioni dei due file LootBot.js e LootBotPlus.js rendono estremamente complicato il merge delle modifiche che nel frattempo vengono caricate sul branch principale, con il rischio di introdurre errori. Ho deciso quindi di procedere per piccole modifiche invece di un refactoring completo, anche se questo comporterà qualche PR in più, in modo da dare la possibilità di iniziare ad adottare i moduli progressivamente. 

Ora che ho più tempo libero riparto, sulla scia di quanto già suggerito da @nrc382 e, prima ancora di mandare la suddivisione in moduli, vorrei procedere con la creazione di una piccola suite di test, per accertarmi di non introdurre errori nelle varie sincronizzazioni con il branch principale.

Volevo chiederti due cose: 

1. LootBot utilizza npm o yarn per la gestione dei pacchetti? nel src code vedo entrambi ma sarebbe sano utilizzarne uno solo o si rischia di avere divergenze tra i due sistemi (nei file lock). 
2. Hai preferenze per il framework di test? pensavo di usare mocha e chai che sono molto diffusi quindi dovrebbero essere familiari anche ad altri sviluppatori e di usare docker/podman per creare le istanze di test (test DB+fake telegram server e client). 

grazie in anticipo!

**Reply by sidelux** [reply]
1. NPM, non mi pare di aver mai usato yarn, forse all'inizio per qualche test
2. Va bene come proponi tu, non ne ho uno preferito

**Reply by furins** [reply]
perfetto! grazie.
In tal caso andrebbe rimosso yarn.lock dalla repo [vedi qui, ad esempio, alcune motivazioni](https://github.com/yarnpkg/yarn/issues/5654)
